### PR TITLE
Feature/serial communication

### PR DESCRIPTION
**Pull request by steersbob**

### CODE REVIEW

File: test/test_serial_comm.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+        for expected in expected_events:
+            actual = c.events.get_nowait()
+            assert actual == expected
+        assert c.events.empty()
+
+        for expected in expected_data:
+            actual = c.data.get_nowait()
+            assert actual == expected
+        assert c.events.empty()
```
Comment (Kargathia) [inline test/test_serial_comm.py]: Assert c.data is empty

File: brewblox_devcon_spark/commander.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            await self._requests[raw_request].put(response)
+        except Exception as ex:
+            LOGGER.error(ex)
+
+    async def _command(self, cmd, **kwargs):
+        raw_request = cmd.request.build(dict(**kwargs))
+        await self._conduit.write_encoded(codecs.encode(raw_request, 'hex'))
+
+        # Wait for a request resolution (matched by request)
+        return await self._requests[raw_request].get()
```
Comment (elcojacobs) [inline brewblox_devcon_spark/commander.py]: Add timeout cleanup for:
- Outstanding requests that received no reply (empty queues) (request timestamp)
- Replies from the controller that are not fetched from the queue (caller already timed out or cancelled request)
- Stale replies from the controller (request should not be resolved with old data) (reply timestamp)

File: brewblox_devcon_spark/communication.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        transport.serial.rts = False
+        LOGGER.debug(f'Serial connection made: {transport}')
+
+    def data_received(self, data):
+        self._buffer += data.decode()
+
+        # Annotations use < and > as start/end characters
+        # Most annotations can be discarded, except for event messages
+        # Event messages are annotations that start with !
+        for event in self._coerce_message_from_buffer(start='<', end='>', filter_expr='!([\s\S]*)'):
```
[the remark this answers is not in the document]
Comment (steersbob) [reply brewblox_devcon_spark/communication.py]: Added `<!interrupted! message>` to the tests, and that was handled as it should be.

File: brewblox_devcon_spark/commands.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        self.header = self.opcode_struct + header
+        self.request = self.header + request
+        self.response = self.header + response
+
+
+def _add_command(opcode, header=Struct(), request=Struct(), response=Struct()):
+    COMMANDS[opcode] = Command(opcode, header, request, response)
+
+
+def identify(unhexed: bytes) -> 'Command':
```
[the remark this answers is not in the document]
Comment (steersbob) [reply brewblox_devcon_spark/commands.py]: Done

File: brewblox_devcon_spark/commands.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        'type' / BrewBloxObjectTypeEnum
+    ),
+    request=Struct(
+        'size' / Default(Int8ub, 0)
+    ),
+    response=Struct(
+        'expectedsize' / Int8sb,
+        'real-type' / BrewBloxObjectTypeEnum,
+        Padding(1),
+        'data' / Optional(PascalString(VarInt, 'utf8')),  # TODO(Bob): should encoding be utf8?
```
[the remark this answers is not in the document]
Comment (steersbob) [reply brewblox_devcon_spark/commands.py]: Replaced Prefixed arrays and pascalstring with `Byte[:]` for unspecified length list.

File: brewblox_devcon_spark/commands.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    header=Struct(
+        'id' / VariableLengthIDAdapter()
+    ),
+    response=Struct(
+        'status' / Int8sb,
+        Terminated
+    )
+)
+
+_add_command(
```
[the remark this answers is not in the document]
Comment (steersbob) [reply brewblox_devcon_spark/commands.py]: `commander._command()` now checks for `response is not None`

File: brewblox_devcon_spark/communication.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (elcojacobs) [reply brewblox_devcon_spark/communication.py]: But what about <should be ignored !this is not a valid event>?

[the remark this answers is not in the document]
Comment (steersbob) [reply brewblox_devcon_spark/communication.py]: That triggered it. with the added ^ `<invalid! event!>` is now handled correctly

**Pull request merged**